quadtree/quadtree.go

```diff
@@ -0,0 +1,385 @@
+	for i := range ps {
+		q.Insert(&ps[i])
+	}
+
+	return q
+}
+
+// NewFromPointers creates a quadtree from a set of pointers.
+func NewFromPointers(points []geo.Pointer) *Quadtree {
+	if len(points) == 0 {
+		// This is kind of meaningless but is what will happen
+		// if using an empty pointset above.
+		return &Quadtree{
+			bound: geo.NewBound(0, 0, 0, 0),
+		}
+	}
+
+	b := geo.NewBoundFromPoints(points[0].Point(), points[0].Point())
+	for _, p := range points {
+		b.Extend(p.Point())
+	}
+
+	q := &Quadtree{
+		bound:     b,
+		freeNodes: make([]node, len(points), len(points)),
+	}
+
+	for _, p := range points {
+		q.Insert(p)
+	}
+
+	return q
+}
+
+// Bound returns the bounds used for the quad tree.
+func (q *Quadtree) Bound() *geo.Bound {
+	return q.bound
+}
+
+// Insert puts an object into the quad tree, must be within the quadtree bounds.
+// If the pointer returns nil, the point will be ignored.
+// This function is not thread-safe, ie. multiple goroutines cannot insert into
+// a single quadtree.
+func (q *Quadtree) Insert(p geo.Pointer) error {
+	if p == nil {
+		return nil
+	}
+
+	point := p.Point()
+	if point == nil {
+		return nil
+	}
+
+	if !q.bound.Contains(point) {
+		return ErrPointOutsideOfBounds
+	}
+
+	if q.root == nil {
+		q.root = &node{}
+	}
+
+	q.insert(q.root, p,
+		q.bound.Left(), q.bound.Right(),
+		q.bound.Bottom(), q.bound.Top(),
+	)
+
+	return nil
+}
+
+// nextNode returns the next node from a preallocated list.
+// This resulted in about 15% improvement in quadtree creation.
+func (q *Quadtree) nextNode() *node {
+	if l := len(q.freeNodes); q.freeIndex >= l {
+		// Exponentially decrease the preallocation size.
+		// On a handful of tests, number of nodes was about 1.5 times pointers.
+		l /= 2
+
+		// min size of the preallocation. I think this could be bigger as it's
+		// not that much memory overhead. Optimizing this more would need
+		// to be use case specific.
+		if l < 25 {
+			l = 25
+		}
+
+		q.freeNodes = make([]node, l, l)
+		q.freeIndex = 0
+	}
+
+	n := &q.freeNodes[q.freeIndex]
+	q.freeIndex++
+	return n
+}
+
+func (q *Quadtree) insert(n *node, p geo.Pointer, left, right, bottom, top float64) {
+	point := p.Point()
+	if n.internal {
+		i := 0
+
+		if cy := (bottom + top) / 2.0; point.Y() <= cy {
+			top = cy
+			i = 2
+		} else {
+			bottom = cy
+		}
+
+		if cx := (left + right) / 2.0; point.X() >= cx {
+			left = cx
+			i++
+		} else {
+			right = cx
+		}
+
+		if n.children[i] == nil {
+			n.children[i] = q.nextNode()
+			n.children[i].pointer = p
+			return
+		}
+
+		q.insert(n.children[i], p, left, right, bottom, top)
+		return
+	}
+
+	if n.pointer == nil {
+		n.pointer = p
+		return
+	}
+
+	// leaf node with a point in it.
+	nPoint := n.pointer.Point()
+
+	// TODO: parameterize this threshold, maybe 1/100th of bound width/height
+	if d := math.Abs(nPoint.X()-point.X()) + math.Abs(nPoint.Y()-point.Y()); d < 0.01 {
+		// similar/duplicate point to stop recursion.
+		n.internal = true
```

Comment:
It seems like in the rest of this function, the node is getting turned into an internal node? Could you move the setting of n.internal outside of this block?